framework/include/constraints/NodeElemConstraintBase.h

```diff
@@ -0,0 +1,121 @@
+#include "NeighborCoupleableMooseVariableDependencyIntermediateInterface.h"
+
+namespace libMesh
+{
+template <typename T>
+class SparseMatrix;
+}
+
+/**
+ * A NodeElemConstraintBase is used when you need to create constraints between
+ * a secondary node and a primary element. It works by allowing you to modify the
+ * residual and jacobian entries on the secondary node and the primary element.
+ */
+class NodeElemConstraintBase
+  : public Constraint,
+    public NeighborCoupleableMooseVariableDependencyIntermediateInterface,
+    public NeighborMooseVariableInterface<Real>
+{
+public:
+  static InputParameters validParams();
+
+  NodeElemConstraintBase(const InputParameters & parameters);
+  ~NodeElemConstraintBase();
+
+  /// Compute the value the secondary node should have at the beginning of a timestep.
+  void computeSecondaryValue(NumericVector<Number> & current_solution);
+
+  /**
+   * Whether or not this constraint should be applied.
+   * @return bool true if this constraint is active, false otherwise
+   */
+  virtual bool shouldApply() { return true; }
+
+  /**
+   * Whether or not the secondary's residual should be overwritten.
+   * @return bool When this returns true the secondary's residual as computed by the constraint will
+   * _replace_ the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSecondaryResidual();
+
+  /**
+   * Whether or not the secondary's Jacobian row should be overwritten.
+   * @return bool When this returns true the secondary's Jacobian row as computed by the constraint
+   * will _replace_ the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSecondaryJacobian() { return overwriteSecondaryResidual(); };
+
+  /**
+   * The variable on the primary elem.
+   * @return MooseVariable & a reference to the primary variable
+   */
+  virtual MooseVariable & primaryVariable() { return _primary_var; }
+
+  /**
+   * The variable number that this object operates on.
+   */
+  const MooseVariable & variable() const override { return _var; }
+
+protected:
+  /// Compute the value the secondary node should have at the beginning of a timestep.
+  virtual Real computeQpSecondaryValue() = 0;
+
+  MooseVariable & _var;
+  /// Primary side variable
+  MooseVariable & _primary_var;
+
+  /// secondary block id
+  unsigned short _secondary;
+  /// primary block id
+  unsigned short _primary;
+
+  /// current node being processed
+  const Node * const & _current_node;
+  const Elem * const & _current_elem;
+
+  /// Shape function on the secondary side.
+  VariablePhiValue _phi_secondary;
+  /// Shape function on the secondary side.  This will always only have one entry and that entry will always be "1"
+  VariableTestValue _test_secondary;
+
+  /// Side shape function.
+  const VariablePhiValue & _phi_primary;
+
+  /// Side test function
+  const VariableTestValue & _test_primary;
+
+  const std::map<dof_id_type, std::vector<dof_id_type>> & _node_to_elem_map;
```

Comment:
docstring
is this a local one or is this the one from the MooseMesh

Reply:
MooseMesh